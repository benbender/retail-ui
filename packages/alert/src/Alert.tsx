import { ButtonIcon, ButtonIconProps } from '@retail-ui/button-icon'
import {
  BanOutline,
  CheckOutline,
  ExclamationCircleOutline,
  ExclamationOutline,
  XSolid,
} from '@retail-ui/heroicons'
import { Theme, useThemeCtx } from '@retail-ui/theme'
import clsx from 'clsx'
import * as React from 'react'

type ReactDivProps = Omit<React.HTMLAttributes<HTMLDivElement>, ''>
type Ref = HTMLDivElement

const ALERT_STATUS_ICON = {
  primary: ExclamationCircleOutline,
  success: CheckOutline,
  danger: BanOutline,
  warning: ExclamationOutline,
}

type AlertVariant = keyof Theme['AlertStyles']['variant']
type AlertColor = keyof Theme['AlertStyles']['variant']['default']
type AlertSize = keyof Theme['AlertStyles']['size']
type AlertBordered = keyof Theme['AlertStyles']['bordered']
type AlertStatusIcon = keyof typeof ALERT_STATUS_ICON

export type AlertProps = {
  content: React.ReactNode
  variant?: AlertVariant
  color?: AlertColor
  statusIcon?: AlertStatusIcon
  size?: AlertSize
  bordered?: AlertBordered
  hasClose?: boolean
  onClose?: () => void
}

export const Alert = React.forwardRef<Ref, ReactDivProps & AlertProps>(
  (props, ref) => {
    const {
      content,
      children,
      variant = 'default',
      color = 'primary',
      size = 'base',
      bordered,
      hasClose = false,
      statusIcon,
      onClose,
      className,
      ...rest
    } = props

    const {
      theme: { AlertStyles },
    } = useThemeCtx()
    const variantCls = AlertStyles.variant

    const colorCls = variantCls[variant]
    const sizeCls = AlertStyles.size
    const borderedCls = AlertStyles.bordered

    const cls = clsx(
      className,
      AlertStyles.base,
      colorCls[color],
      sizeCls[size],
      bordered && borderedCls[bordered],
      hasClose && AlertStyles.hasClose,
    )

    const btnCls = clsx(AlertStyles.btnClose)
    const btnVariant: Record<AlertVariant, ButtonIconProps['variant']> = {
      default: 'light',
      outline: 'outline',
      solid: 'default',
    }

    const StatusIcon = statusIcon ? ALERT_STATUS_ICON[statusIcon] : null
    const statusIconCls = clsx(AlertStyles.statusIcon)

    const descriptionCls = clsx(AlertStyles.description)

    return (
      <div ref={ref} className={cls} role="alert" {...rest}>
        {StatusIcon && (
          <StatusIcon
            className={statusIconCls}
            // style={{
            //   maxWidth: '1.25rem',
            //   minWidth: '1.25rem',
            //   maxHeight: '1.25rem',
            //   minHeight: '1.25rem',
            // }}
          />
        )}
        {!children && content}
        {children && (
          <div>
            <span>{content}</span>
            <p className={descriptionCls}>{children}</p>
          </div>
        )}
        {hasClose && (
          <div className={btnCls}>
            <ButtonIcon
              color={color}
              size="base"
              shape="rounded"
              icon={XSolid}
              variant={btnVariant[variant]}
              onClick={onClose}
            />
          </div>
        )}
      </div>
    )
  },
)

Alert.displayName = 'Alert'
