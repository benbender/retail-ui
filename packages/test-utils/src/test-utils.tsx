import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
