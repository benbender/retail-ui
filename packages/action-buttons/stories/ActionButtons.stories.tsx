import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
