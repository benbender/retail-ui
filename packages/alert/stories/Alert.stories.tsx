import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
