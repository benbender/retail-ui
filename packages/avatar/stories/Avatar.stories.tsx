import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
