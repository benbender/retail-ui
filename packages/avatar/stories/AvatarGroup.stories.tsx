import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
