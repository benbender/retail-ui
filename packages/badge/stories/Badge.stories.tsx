import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
