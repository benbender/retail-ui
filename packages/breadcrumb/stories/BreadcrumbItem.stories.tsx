import * as React from 'react'

import { Breadcrumb, BreadcrumbItem, BreadcrumbProps } from '../src'

const meta = {
  title: 'Retail-UI/BreadcrumbItem',
  component: BreadcrumbItem,
}

export default meta

const Template = (args: BreadcrumbProps) => (
  <Breadcrumb data-testid={meta.title}>
    <BreadcrumbItem {...args}>
      <a href="#">Dashboard</a>
    </BreadcrumbItem>
    <BreadcrumbItem {...args}>
      <a href="#">Project</a>
    </BreadcrumbItem>
    <BreadcrumbItem isActive {...args}>
      Export
    </BreadcrumbItem>
  </Breadcrumb>
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Breadcrumb',
}
