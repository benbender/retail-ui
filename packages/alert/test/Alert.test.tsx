import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Alert from '../stories/Alert.stories'

describe('Alert', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Alert.Simple content="Test title" />)
    expect(asFragment()).toMatchSnapshot()
  })
})
