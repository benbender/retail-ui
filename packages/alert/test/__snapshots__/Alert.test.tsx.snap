// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders correctly 1`] = `
<DocumentFragment>
  <div
    class="font-medium relative overflow-hidden border-2 border-solid rounded-md duration-150 ease-in flex items-center justify-start bg-purple-200 border-purple-200 text-purple-800 py-2 pl-4 pr-4 text-base"
    data-testid="Retail-UI/Alert"
    role="alert"
  >
    Test title
  </div>
</DocumentFragment>
`;
